File: 1371-minimum-remove-to-make-valid-parentheses/minimum-remove-to-make-valid-parentheses.cs
public class Solution {
    public string MinRemoveToMakeValid(string s) {
        Stack<int> stack = new Stack<int>();
        HashSet<int> toRemove = new HashSet<int>();

        for(int i = 0; i < s.Length; i++)
        {
            if(s[i] == '(')
            {
                stack.Push(i);
            }
            else if(s[i] == ')')
            {
                if(stack.Count > 0)
                {
                    stack.Pop();
                }
                else{
                    toRemove.Add(i);
                }
            }
        }

        while(stack.Count > 0)
        {
            toRemove.Add(stack.Pop());
        }

        StringBuilder result = new StringBuilder();
        for(int i = 0; i < s.Length; i++)
        {
            if(!toRemove.Contains(i))
            {
                result.Append(s[i]);
            }
        }

        return result.ToString();

    }
}

File: 1371-minimum-remove-to-make-valid-parentheses/minimum-remove-to-make-valid-parentheses.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        vector<char> res;
        int count = 0;

        for (char c : s) {
            if (c == '(') {
                res.push_back(c);
                count += 1;
            }
            else if (c == ')' && count > 0) {
                res.push_back(c);
                count -= 1;
            }
            else if (c != ')') {
                res.push_back(c);
            }
        }

        string str = "";
        for (int i = res.size() - 1; i >= 0; i--) {
            if (res[i] == '(' && count > 0) {
                count -= 1;
            } else {
                str += res[i];  
            }
        }

        reverse(str.begin(), str.end());
        return str;
    }
};